File: web/src/test/java/no/nav/modiapersonoversikt/service/unleash/strategier/ByEnvironmentStrategyTest.java
package no.nav.modiapersonoversikt.service.unleash.strategier;

import org.junit.jupiter.api.Test;

import java.util.HashMap;

import static no.nav.common.test.SystemProperties.setTemporaryProperty;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.is;

class ByEnvironmentStrategyTest {


    private final ByEnvironmentStrategy strategy = new ByEnvironmentStrategy();
    private final String ENVIRONMENT_PROPERTY = "APP_ENVIRONMENT_NAME";

    @Test
    void environmentTest() {
        assertThat(strategy.isEnabled(null), is(false));

        assertEnabled("t6", "t6");
        assertDisabled("p", "q0");
        assertDisabled(null, "q0");
        assertDisabled(null, "");
        assertDisabled("q0", null);
        assertDisabled("", null);
        assertDisabled("", "");
        assertDisabled("", ",,,");

        assertEnabled("q1", "q6,q1,t6");
    }

    private void assertEnabled(String actualEnvironment, String environmentParameter) {
        assertEnvironment(actualEnvironment, environmentParameter, true);
    }

    private void assertDisabled(String actualEnvironment, String environmentParameter) {
        assertEnvironment(actualEnvironment, environmentParameter, false);
    }

    private void assertEnvironment(String actualEnvironment, String toggleParameter, boolean expectedStatus){
        setTemporaryProperty(ENVIRONMENT_PROPERTY, actualEnvironment, () -> {
            HashMap<String, String > params = new HashMap<>() {{
                put("miljø", toggleParameter);
            }};
            assertThat(strategy.isEnabled(params), is(expectedStatus));
        });
    }
}

File: web/src/test/java/no/nav/modiapersonoversikt/service/unleash/strategier/ByEnvironmentStrategyTest.kt
package no.nav.modiapersonoversikt.service.unleash.strategier

import no.nav.modiapersonoversikt.service.unleash.strategier.StrategyUtils.ENVIRONMENT_PROPERTY
import no.nav.personoversikt.common.test.testenvironment.TestEnvironment.Companion.withEnvironment
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource

internal class ByEnvironmentStrategyTest {
    val byEnvironmentStrategy = ByEnvironmentStrategy()

    @ParameterizedTest
    @MethodSource("testcases")
    internal fun `should evaluate strategy`(
        unleashConfig: String?,
        simulatedEnv: String?,
        expected: Boolean,
    ) {
        withEnvironment(mapOf(ENVIRONMENT_PROPERTY to simulatedEnv)) {
            val parameters = mapOf(ByEnvironmentStrategy.ENABLED_ENVIRONMENT_PROPERTY to unleashConfig)

            assertThat(byEnvironmentStrategy.isEnabled(parameters)).isEqualTo(expected)
        }
    }

    companion object {
        @JvmStatic
        fun testcases() =
            listOf(
                strategyEnabled(simulatedEnv = "t6", unleashConfig = "t6"),
                strategyEnabled(simulatedEnv = "q1", unleashConfig = "q6,q1,t6"),
                strategyDisabled(simulatedEnv = "p", unleashConfig = "q0"),
                strategyDisabled(simulatedEnv = null, unleashConfig = "q0"),
                strategyDisabled(simulatedEnv = null, unleashConfig = ""),
                strategyDisabled(simulatedEnv = "q0", unleashConfig = null),
                strategyDisabled(simulatedEnv = "", unleashConfig = null),
                strategyDisabled(simulatedEnv = "", unleashConfig = ""),
                strategyDisabled(simulatedEnv = "", unleashConfig = ",,,"),
            )

        private fun strategyEnabled(
            unleashConfig: String?,
            simulatedEnv: String?,
        ): Arguments {
            return Arguments.of(unleashConfig, simulatedEnv, true)
        }

        private fun strategyDisabled(
            unleashConfig: String?,
            simulatedEnv: String?,
        ): Arguments {
            return Arguments.of(unleashConfig, simulatedEnv, false)
        }
    }
}
